### ISSUE TITLE
Bug: Goroutine leak in `coderd.(*api).workspaceAgentTurn`

### ISSUE THREAD

**Issue by mafredri** [coder/coder#1508]
There seems to be a goroutine leak in `coderd.(*api).workspaceAgentTurn`.

This could be seen as two bugs:

1. Goroutine leak
2. Use of `(*http.Request).Context()` after `Hijack` in more than one place

## Steps to Reproduce

0. Enable pprof for coder server
1. `coder ssh dev`
3. `ctlr+d`
4. Goto 1
5. Check pprof (`go tool pprof -http=:8080 http://localhost:6060/debug/pprof/goroutine`)

The leak is in part due to reliance on the `http.Request` context and use of websockets. The underlying websocket library calls `(*http.Request).Hijack` which disables context propagation.

This happens here: https://github.com/coder/coder/blob/668a6712e62a839a193544e431552744520b05df/coderd/workspaceagents.go#L304

And the following contexts will not cancel until the http handler completes:
https://github.com/coder/coder/blob/668a6712e62a839a193544e431552744520b05df/coderd/workspaceagents.go#L316

https://github.com/coder/coder/blob/668a6712e62a839a193544e431552744520b05df/coderd/workspaceagents.go#L320

We must avoid using `r.Context()` after hijack, unless we are using it with the expectation that the http handler will exit (at which point the context will complete).

I'm unfamiliar with the [`pion/turn`](https://github.com/pion/turn) package, but another factor could be wrt how it handles connection closure, perhaps it does not propagate as we expect since we're not calling `wsConn.Close()` due to context reliance?

---

Similar reliance on request context after hijack is done elsewhere, we should rethink all of them. Example:

https://github.com/coder/coder/blob/668a6712e62a839a193544e431552744520b05df/coderd/workspaceagents.go#L155

**Comment by tjcran** [comment]
@mafredri what is the potential impact of this bug?

**Comment by ketang** [comment]
Very interested in how you found this. I'm sure there was more to it than `pprof`.

**Comment by mafredri** [comment]
@tjcran For the codepath I analyzed, I estimate ~0.25MB of memory leaked per SSH connection, for active long running servers it could mean significant memory consumption that is unrecoverable without a restart of the coder server. There could be other potential issues I haven't explored, like file descriptor exhaustion on the host.

@ketang Not terribly exiting I'm afraid, I was simply curious as to what was causing slight CPU usage while `coder server` was idling, so I took a look and found this instead. 😄
